kauri/kauri.go

```diff
@@ -0,0 +1,320 @@
+		k.nodes[hotstuff.ID(n.ID())] = n
+	}
+	k.tree = CreateTree(k.configuration.Len(), k.opts.ID())
+
+	// pIDs := make(map[hotstuff.ID]int)
+	// for id := range k.configuration.ActiveReplicas() {
+	// 	pIDs[id] = index
+	// 	index++
+	// }
+	idMappings := make(map[hotstuff.ID]int)
+	for i := 0; i < k.configuration.Len(); i++ {
+		idMappings[hotstuff.ID(i+1)] = i
+	}
+	k.tree.InitializeWithPIDs(idMappings)
+	k.initDone = true
+	k.senders = make([]hotstuff.ID, 0)
+}
+
+// Begin starts dissemination of proposal and aggregation of votes.
+func (k *Kauri) Begin(pc hotstuff.PartialCert, p hotstuff.ProposeMsg) {
+	if !k.initDone {
+		k.eventLoop.DelayUntil(backend.ConnectedEvent{}, func() { k.Begin(pc, p) })
+		return
+	}
+	k.reset()
+	k.blockHash = pc.BlockHash()
+	k.currentView = p.Block.View()
+	k.aggregatedContribution = pc.Signature()
+	ids := k.randomizeIDS(k.blockHash, k.leaderRotation.GetLeader(k.currentView))
+	k.tree.InitializeWithPIDs(ids)
+	k.SendProposalToChildren(p)
+	waitTime := time.Duration(uint64(k.tree.GetHeight() * 20 * int(time.Millisecond)))
+	go k.aggregateAndSend(waitTime, k.currentView)
+}
+
+func (k *Kauri) reset() {
+	k.aggregatedContribution = nil
+	k.senders = make([]hotstuff.ID, 0)
+	k.isAggregationSent = false
+}
+
+func (k *Kauri) aggregateAndSend(t time.Duration, view hotstuff.View) {
+	ticker := time.NewTicker(t)
+	<-ticker.C
+	ticker.Stop()
+	if k.currentView != view {
+		return
+	}
+	if !k.isAggregationSent {
+		k.SendContributionToParent()
+	}
+}
+
+// SendProposalToChildren sends the proposal to the children
+func (k *Kauri) SendProposalToChildren(p hotstuff.ProposeMsg) {
+	children := k.tree.GetChildren()
+	if len(children) != 0 {
+		config, err := k.configuration.SubConfig(k.tree.GetChildren())
+		if err != nil {
+			k.logger.Error("Unable to send the proposal to children", err)
+			return
+		}
+		k.logger.Debug("sending proposal to children ", k.tree.GetChildren())
+		config.Propose(p)
+	} else {
+		k.SendContributionToParent()
+		k.isAggregationSent = true
+	}
+}
+
+// OnContributionRecv is invoked upon receiving the vote for aggregation.
+func (k *Kauri) OnContributionRecv(event ContributionRecvEvent) {
+	if k.currentView != hotstuff.View(event.Contribution.View) {
+		return
+	}
+	contribution := event.Contribution
+	k.logger.Debug("processing the contribution from ", contribution.ID)
+	currentSignature := hotstuffpb.QuorumSignatureFromProto(contribution.Signature)
+	_, err := k.mergeWithContribution(currentSignature)
+	if err != nil {
+		k.logger.Debug("Unable to merge the contribution from ", contribution.ID)
+		return
+	}
+	k.senders = append(k.senders, hotstuff.ID(contribution.ID))
+	if isSubSet(k.tree.GetSubTreeNodes(), k.senders) {
+		k.SendContributionToParent()
+		k.isAggregationSent = true
+	}
+}
+
+// SendContributionToParent sends contribution to the parent node.
+func (k *Kauri) SendContributionToParent() {
+	parent, ok := k.tree.GetParent()
+	if ok {
+		node, isPresent := k.nodes[parent]
```

Comment:
Consider moving the `nodes` map to the tree and let `tree.Parent()` return the parent node directly instead of the `hotstuff.ID`.